tensorflow/core/kernels/resize_bilinear_op_gpu.cu.cc

```diff
@@ -338,6 +389,55 @@ __global__ void LegacyResizeBilinearGradKernel(
   }
 }
 
+template <typename T>
+__global__ void LegacyResizeBilinearDeterministicGradKernel(
```

Comment:
Is this the same kernel as `ResizeBilinearDeterministicGradKernel`, except in_x/y_start? If so, would it make sense to share the kernel code and pass in the pixel offset?

Reply:
Yes, this is the same code, except for the calculations of `in_y_start`, `out_y_start`, in_x_start`, and `out_y_start`. I followed the existing pattern of replicating the code, as set by `ResizeBilinearGradKernel` and `LegacyResizeBilinearGradKernel`.

What you're suggesting is an improvement. I'm considering changing it for those existing, non-deterministic kernels as well. What do you think?

Reply:
I just merged the legacy and non-legacy deterministic kernels (now running tests locally) and note that this makes the legacy kernel slightly slower than before, not only because of having to add in the zero offsets, but also because the legacy kernel skipped an unnecessary clamping operation. That clamping operation always runs in the merged kernel. I still think it's better to have one kernel.

I still think it's better to be factored like this.

Further re-factoring, if its done, will be submitted in a separate PR.

Reply:
Commit pushed.

Reply:
I also just looked into combining the existing, non-deterministic back-prop kernels into a single kernel with an offset parameter. It think it's doable and probably worth it. I won't submit another pull request for that right now because it will collide with this current pull request.

I think it's also possible to combine two of the forward kernels into a single kernel with an offset parameter. That could probably be done in the same future pull request.